pkg/uprobetracer/tracer.go

```diff
@@ -182,30 +197,66 @@ func (t *Tracer[Event]) searchForLibrary(containerPid uint32) ([]string, error)
 	return ldCachePaths, nil
 }
 
+func fillArray[T any](generator func() T, count int) []T {
+	result := make([]T, count)
+	for i := 0; i < count; i++ {
+		result[i] = generator()
+	}
+	return result
+}
+
 // attach uprobe program to the inode of the file passed in parameter
-func (t *Tracer[Event]) attachUprobe(file *os.File) (link.Link, error) {
+func (t *Tracer[Event]) attachUprobe(file *os.File) (link.Link, *func(), error) {
 	attachPath := path.Join(host.HostProcFs, "self/fd/", fmt.Sprint(file.Fd()))
 	ex, err := link.OpenExecutable(attachPath)
 	if err != nil {
-		return nil, fmt.Errorf("opening %q: %w", attachPath, err)
+		return nil, nil, fmt.Errorf("opening %q: %w", attachPath, err)
 	}
 	switch t.progType {
 	case ProgUprobe:
-		return ex.Uprobe(t.attachSymbol, t.prog, nil)
+		link, err := ex.Uprobe(t.attachSymbol, t.prog, nil)
+		return link, nil, err
 	case ProgUretprobe:
-		return ex.Uretprobe(t.attachSymbol, t.prog, nil)
+		link, err := ex.Uretprobe(t.attachSymbol, t.prog, nil)
+		return link, nil, err
 	case ProgUSDT:
-		attachInfo, err := getUsdtInfo(attachPath, t.attachSymbol)
+		attachInfo, err := t.getUsdtInfo(attachPath, t.attachSymbol)
 		if err != nil {
-			return nil, fmt.Errorf("reading USDT metadata: %w", err)
+			return nil, nil, fmt.Errorf("reading USDT metadata: %w", err)
 		}
-		return ex.Uprobe(t.attachSymbol, t.prog,
-			&link.UprobeOptions{
-				Address:      attachInfo.attachAddress,
-				RefCtrOffset: attachInfo.semaphoreAddress,
+
+		length := len(attachInfo.attachAddresses)
+		cookies := fillArray[uint64](func() uint64 {
+			// Since an ELF file may have multiple USDT trace points,
+			// we generate an unique ID to avoid conflict
+			return usdtCookieUUID.Add(1)
+		}, length)
+
+		l, err := ex.UprobeMulti(
+			fillArray[string](func() string {
+				return t.attachSymbol
+			}, length),
+			t.prog,
+			&link.UprobeMultiOptions{
+				Addresses:     attachInfo.attachAddresses,
+				RefCtrOffsets: attachInfo.semaphoreAddresses,
+				Cookies:       cookies,
 			})
+		if err != nil {
+			return nil, nil, err
+		}
+
+		for i := 0; i < length; i++ {
+			usdtArgsInfoMap.Update(cookies[i], attachInfo.arguments[i], ebpf.UpdateAny)
```

Comment:
Will `NoExists` bring better performance? Here I just want to write a value into the map. I thought if there exists some unknown reasons that cause the old values still being in the map, we can still use `UpdateAny` to overwrite them, and it will work well.